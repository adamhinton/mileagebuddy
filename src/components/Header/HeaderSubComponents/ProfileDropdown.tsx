// ____________________________________________________________________
// This is the dropdown that appears when user clicks their profile icon in Header.tsx
// ____________________________________________________________________

import { type User } from "@/app/zod/schemas/UserSchema";

type ProfileDropdownProps = {
	user: User;
	isProfileDropdownOpen: boolean;
	setIsProfileDropdownOpen: (isOpen: boolean) => void;
	handleSignOut: () => void;
	profileDropdownRef: React.RefObject<HTMLDivElement>;
};

const ProfileDropdown = (props: ProfileDropdownProps) => {
	const {
		user,
		isProfileDropdownOpen,
		setIsProfileDropdownOpen,
		handleSignOut,
		profileDropdownRef,
	} = props;

	return (
		<div className="relative" ref={profileDropdownRef}>
			{/* Profile button - Icon on small screens, full text on md+ screens */}
			<button
				onClick={() => setIsProfileDropdownOpen(!isProfileDropdownOpen)}
				className="flex items-center px-3 py-1 bg-background-elevated rounded-full text-sm text-neutral-text border border-primary-50 dark:border-primary-200 hover:bg-background-highlight transition-colors"
			>
				{/* Always visible profile icon */}
				<div className="w-2 h-2 rounded-full bg-accent mr-2 animate-pulse"></div>

				{/* TODO replace with profile icon */}
				<span className="hidden sm:inline" data-testid="profile-icon">
					Profile
				</span>

				{/* Dropdown indicator */}
				<svg
					xmlns="http://www.w3.org/2000/svg"
					className="h-4 w-4 ml-2"
					fill="none"
					viewBox="0 0 24 24"
					stroke="currentColor"
				>
					<path
						strokeLinecap="round"
						strokeLinejoin="round"
						strokeWidth={2}
						d="M19 9l-7 7-7-7"
					/>
				</svg>
			</button>

			{/* Profile dropdown menu */}
			{isProfileDropdownOpen && (
				<div className="absolute right-0 mt-2 w-48 bg-background-elevated rounded-md shadow-lg border border-primary-50 dark:border-primary-200 z-40">
					<div className="p-2">
						<div className="text-sm font-medium text-neutral-text mb-2 border-b border-primary-50 dark:border-primary-200 pb-2">
							{user?.email}
						</div>
						<button
							onClick={handleSignOut}
							className="w-full text-left px-3 py-2 text-sm text-red-500 hover:bg-red-50 dark:hover:bg-red-900/20 rounded-md transition-colors"
						>
							Sign Out
						</button>
					</div>
				</div>
			)}
		</div>
	);
};

export default ProfileDropdown;
