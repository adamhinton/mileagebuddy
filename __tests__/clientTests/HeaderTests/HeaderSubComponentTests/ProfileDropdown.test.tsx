// _______________________________________________________________
// This is (obviously) tests for ProfileDropdown.tsx
// The component is pretty simple, so not much to test here
// _______________________________________________________________

import ProfileDropdown from "@/components/Header/HeaderSubComponents/ProfileDropdown";
import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

const createMockProps = (overrides = {}) => ({
	user: { id: "test-user-id", email: "test@example.com", isDarkMode: false },
	isProfileDropdownOpen: false,
	setIsProfileDropdownOpen: jest.fn(),
	handleSignOut: jest.fn(),
	profileDropdownRef: { current: document.createElement("div") },
	...overrides,
});

describe("ProfileDropdown.tsx", () => {
	it("Renders without errors", () => {
		const mockProps = createMockProps();
		render(<ProfileDropdown {...mockProps} />);
	});

	describe("Dropdown Behavior", () => {
		it("Does not show dropdown menu when closed", () => {
			const mockProps = createMockProps({ isProfileDropdownOpen: false });
			const { queryByText } = render(<ProfileDropdown {...mockProps} />);

			const emailElement = queryByText("test@example.com");
			expect(emailElement).not.toBeInTheDocument();

			// The Sign Out button should not be visible
			const signOutButton = queryByText(/sign out/i);
			expect(signOutButton).not.toBeInTheDocument();
		});

		it("Shows dropdown menu when open", () => {
			const mockProps = createMockProps({ isProfileDropdownOpen: true });
			const { getByText } = render(<ProfileDropdown {...mockProps} />);

			const dropdownMenu =
				getByText("test@example.com").closest("div.absolute");
			expect(dropdownMenu).toBeInTheDocument();
			expect(dropdownMenu).toBeVisible();
		});

		it("Displays correct user email in the dropdown", () => {
			const customEmail = "custom.user@test.com";
			const mockProps = createMockProps({
				isProfileDropdownOpen: true,
				user: { id: "test-id", email: customEmail, isDarkMode: false },
			});

			const { getByText } = render(<ProfileDropdown {...mockProps} />);

			const emailElement = getByText(customEmail);
			expect(emailElement).toBeInTheDocument();
			expect(emailElement).toBeVisible();
		});
	});

	describe("Interaction", () => {
		it("Toggles dropdown when profile button is clicked", () => {
			const setIsProfileDropdownOpenMock = jest.fn();
			const mockProps = createMockProps({
				isProfileDropdownOpen: false,
				setIsProfileDropdownOpen: setIsProfileDropdownOpenMock,
			});

			render(<ProfileDropdown {...mockProps} />);

			const profileButton = screen.getByTestId("profile-icon");
			fireEvent.click(profileButton);

			expect(setIsProfileDropdownOpenMock).toHaveBeenCalledWith(true);

			const openMockProps = createMockProps({
				isProfileDropdownOpen: true,
				setIsProfileDropdownOpen: setIsProfileDropdownOpenMock,
			});

			setIsProfileDropdownOpenMock.mockReset();

			render(<ProfileDropdown {...openMockProps} />);

			fireEvent.click(profileButton);
		});

		it("Calls handleSignOut when sign out button is clicked", () => {
			// Setup with dropdown open
			const handleSignOutMock = jest.fn();
			const mockProps = createMockProps({
				isProfileDropdownOpen: true,
				handleSignOut: handleSignOutMock,
			});

			const { getByRole } = render(<ProfileDropdown {...mockProps} />);

			const signOutButton = getByRole("button", { name: /sign out/i });
			fireEvent.click(signOutButton);

			// Assert that handleSignOut was called
			expect(handleSignOutMock).toHaveBeenCalled();
		});
	});
});
