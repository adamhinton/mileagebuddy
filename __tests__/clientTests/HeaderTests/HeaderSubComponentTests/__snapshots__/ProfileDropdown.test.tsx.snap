// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileDropdown.tsx Snapshot tests matches snapshot with custom email from 4.30.2025 1`] = `
<div>
  <div
    class="relative"
  >
    <button
      class="flex items-center px-3 py-1 bg-background-elevated rounded-full text-sm text-neutral-text border border-primary-50 dark:border-primary-200 hover:bg-background-highlight transition-colors"
    >
      <div
        class="w-2 h-2 rounded-full bg-accent mr-2 animate-pulse"
      />
      <span
        class="hidden sm:inline"
        data-testid="profile-icon"
      >
        Profile
      </span>
      <svg
        class="h-4 w-4 ml-2"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
    <div
      class="absolute right-0 mt-2 w-48 bg-background-elevated rounded-md shadow-lg border border-primary-50 dark:border-primary-200 z-40"
    >
      <div
        class="p-2"
      >
        <div
          class="text-sm font-medium text-neutral-text mb-2 border-b border-primary-50 dark:border-primary-200 pb-2"
        >
          custom.user@test.com
        </div>
        <button
          class="w-full text-left px-3 py-2 text-sm text-red-500 hover:bg-red-50 dark:hover:bg-red-900/20 rounded-md transition-colors"
        >
          Sign Out
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ProfileDropdown.tsx Snapshot tests matches snapshot with dropdown closed from 4.30.2025 1`] = `
<div>
  <div
    class="relative"
  >
    <button
      class="flex items-center px-3 py-1 bg-background-elevated rounded-full text-sm text-neutral-text border border-primary-50 dark:border-primary-200 hover:bg-background-highlight transition-colors"
    >
      <div
        class="w-2 h-2 rounded-full bg-accent mr-2 animate-pulse"
      />
      <span
        class="hidden sm:inline"
        data-testid="profile-icon"
      >
        Profile
      </span>
      <svg
        class="h-4 w-4 ml-2"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`ProfileDropdown.tsx Snapshot tests matches snapshot with dropdown open from 4.30.2025 1`] = `
<div>
  <div
    class="relative"
  >
    <button
      class="flex items-center px-3 py-1 bg-background-elevated rounded-full text-sm text-neutral-text border border-primary-50 dark:border-primary-200 hover:bg-background-highlight transition-colors"
    >
      <div
        class="w-2 h-2 rounded-full bg-accent mr-2 animate-pulse"
      />
      <span
        class="hidden sm:inline"
        data-testid="profile-icon"
      >
        Profile
      </span>
      <svg
        class="h-4 w-4 ml-2"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
    <div
      class="absolute right-0 mt-2 w-48 bg-background-elevated rounded-md shadow-lg border border-primary-50 dark:border-primary-200 z-40"
    >
      <div
        class="p-2"
      >
        <div
          class="text-sm font-medium text-neutral-text mb-2 border-b border-primary-50 dark:border-primary-200 pb-2"
        >
          test@example.com
        </div>
        <button
          class="w-full text-left px-3 py-2 text-sm text-red-500 hover:bg-red-50 dark:hover:bg-red-900/20 rounded-md transition-colors"
        >
          Sign Out
        </button>
      </div>
    </div>
  </div>
</div>
`;
